src/Microsoft.ApplicationInsights.AspNetCore/TelemetryInitializers/DomainNameRoleInstanceTelemetryInitializer.cs

```diff
@@ -0,0 +1,54 @@
+﻿namespace Microsoft.ApplicationInsights.AspNetCore.TelemetryInitializers
+{
+    using System;
+    using System.Globalization;
+    using System.Net;
+    using System.Net.NetworkInformation;
+    using System.Threading;
+    using Channel;    
+    using Microsoft.ApplicationInsights.Extensibility;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;    
+
+    /// <summary>
+    /// A telemetry initializer that populates cloud context role instance.
+    /// </summary>
+    public class DomainNameRoleInstanceTelemetryInitializer : ITelemetryInitializer
+    {
+        private string roleInstanceName;
+
+        /// <summary>
+        /// Initializes role instance name and node name with the host name.
+        /// </summary>        
+        /// <param name="telemetry">Telemetry item.</param>
+        public void Initialize(ITelemetry telemetry)
+        {
+            if (string.IsNullOrEmpty(telemetry.Context.Cloud.RoleInstance))
+            {
+                var name = LazyInitializer.EnsureInitialized(ref this.roleInstanceName, this.GetMachineName);
+                telemetry.Context.Cloud.RoleInstance = name;
+            }
+
+            InternalContext internalContext = telemetry.Context.GetInternalContext();
```

Comment:
do not set nodeName if you set roleInstance. No need.

CC: @iusafaro

Reply:
No, `nodename` will be set by `RoleInstance` on back end if not set explicitly. Only reason to set it on the environments that has `RoleInstance` name unstable and we want something more constant for billing.

Also nodename is only for legacy customers with existing contracts. No new customers benefit from having nodename. Thus it's ideal to not send it when possible to minimize the footprint